src/syntax.ml

```diff
@@ -1067,6 +1139,330 @@ and explicit_list_items :
+   read. The only "valid" way to exit is by reading a [`Right_brace] token,
+   which is consumed.
+
+   Whitespace inside the list, but outside list items, is not significant – this
+   parsing function consumes all of it. Otherwise, only table cell start tokens
+   are accepted. Everything else is an error. *)
+and heavy_table_row ~parent_markup input =
+  let rec consume_cell_items acc =
+    let next_token = peek input in
+    match next_token.value with
+    | `Right_brace ->
+        junk input;
+        (List.rev acc, next_token.location)
+    | `Space _ | `Single_newline _ | `Blank_line _ ->
+        junk input;
+        consume_cell_items acc
+    | `Begin_table_header as token -> (
+        junk input;
+        (match (peek input).value with
+        | `Space _ | `Single_newline _ | `Blank_line _ | `Right_brace
+        | `Begin_paragraph_style _ ->
+            ()
+        | _ ->
+            Parse_error.should_be_followed_by_whitespace next_token.location
+              ~what:(Token.print token)
+            |> add_warning input);
+        let content, _brace_location =
+          heavy_table_header input ~parent_markup:token
+        in
+        match content with
+        | None -> consume_cell_items ((`Header None, []) :: acc)
+        | Some (x, b) -> consume_cell_items ((`Header x, b) :: acc))
+    | `Begin_table_data as token ->
+        junk input;
+        (match (peek input).value with
+        | `Space _ | `Single_newline _ | `Blank_line _ | `Right_brace -> ()
+        | _ ->
+            Parse_error.should_be_followed_by_whitespace next_token.location
+              ~what:(Token.print token)
+            |> add_warning input);
+        let content, token_after_list_item, _where_in_line =
+          block_element_list In_table_cell ~parent_markup:token input
+        in
+        (match token_after_list_item.value with
+        | `Right_brace -> junk input
+        | `End ->
+            Parse_error.not_allowed token_after_list_item.location
+              ~what:(Token.describe `End) ~in_what:(Token.describe token)
+            |> add_warning input);
+        consume_cell_items ((`Data, content) :: acc)
+    | token ->
+        let suggestion =
+          match token with
+          | `Begin_section_heading _ | `Tag _ | `Begin_table _ ->
+              Printf.sprintf "move %s outside the table." (Token.describe token)
+          | `Begin_table_row ->
+              Printf.sprintf "move %s outside the row item."
+                (Token.describe token)
+          | _ ->
+              Printf.sprintf "move %s into a cell item." (Token.describe token)
+        in
+        Parse_error.not_allowed next_token.location ~what:(Token.describe token)
+          ~in_what:(Token.describe parent_markup)
+          ~suggestion
+        |> add_warning input;
+        junk input;
+        consume_cell_items acc
+  in
+  consume_cell_items []
+
+and heavy_table_header ~parent_markup input =
```

Comment:
The 3 functions have very similar code: Parsing nested-block-elements until a `` `Right_brace ``. The only change is how to handle `` `Begin_table_header | `Begin_table_data | `Begin_paragraph_style `` etc.. I think a more general higher order function is possible.

Reply:
I factorized a bit and removed some unnecessary warnings when parsing.